### PR TITLE
Add new resource to support IAM custom project roles.

### PR DESCRIPTION
**Pull request by rosbo**
Addresses the first half of #540 

I also had to update the IAM client to the latest version.

I wasn't sure if I should name the resource `google_project_iam_role` or `google_project_custom_iam_role`.

In the [general docs](https://cloud.google.com/iam/docs/understanding-custom-roles), it is call `Customized Role` or `Custom Role`. In the [API documentation](https://cloud.google.com/iam/reference/rest/v1/organizations.roles/create), it is only role.

### CODE REVIEW

**Comment by rosbo** [comment]
```sh
TF_ACC=1 go test ./google -v -run TestAccGoogleProjectIamRole_ -timeout 120m
=== RUN   TestAccGoogleProjectIamRole_import
=== RUN   TestAccGoogleProjectIamRole_basic
=== RUN   TestAccGoogleProjectIamRole_undelete
--- PASS: TestAccGoogleProjectIamRole_import (2.86s)
--- PASS: TestAccGoogleProjectIamRole_basic (3.76s)
--- PASS: TestAccGoogleProjectIamRole_undelete (4.86s)
PASS
```

**Comment by pdecat** [comment]
Awesome!

Do you plan to also add `google_organization_iam_role` in the same PR?

**Comment by rosbo** [comment]
@pdecat I will add support for the organization custom IAM roles in a separate PR to make it easier to review.

File: google/resource_google_project_iam_role_test.go [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+			return fmt.Errorf("Incorrect deletion status. Expected %t, got %t", deleted, role.Deleted)
+		}
+
+		return nil
+	}
+}
+
+func testAccCheckGoogleProjectIamRole_basic(roleId string) string {
+	return fmt.Sprintf(`
+resource "google_project_iam_role" "foo" {
```
Comment (danawillow) [inline google/resource_google_project_iam_role_test.go]: can you `terraform fmt` these?

File: website/docs/r/google_project_iam_role.html.markdown [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+[the official documentation](https://cloud.google.com/iam/docs/understanding-custom-roles)
+and
+[API](https://cloud.google.com/iam/reference/rest/v1/organizations.roles).
+
+## Example Usage
+
+This snippet creates a customized IAM role.
+
+```hcl
+resource "google_project_iam_role" "myCustomRole" {
```
Comment (danawillow) [inline website/docs/r/google_project_iam_role.html.markdown]: nit: I don't think we _actually_ have an opinion on this, but I feel like I very rarely see camelcasing in resource names. Most of the examples I found were one word, or used hyphens.

File: website/docs/r/google_project_iam_role.html.markdown [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+  role_id = "myCustomRole"
```
Comment (danawillow) [inline website/docs/r/google_project_iam_role.html.markdown]: terraform fmt

File: google/resource_google_project_iam_role.go [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+			"role_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"title": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"permissions": {
```
Comment (danawillow) [inline google/resource_google_project_iam_role.go]: Should this be `MinItems: 1`?

File: google/resource_google_project_iam_role.go [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+}
+
+func resourceGoogleProjectIamRoleCreate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	project, err := getProject(d, config)
+	if err != nil {
+		return err
+	}
+
```
Comment (danawillow) [inline google/resource_google_project_iam_role.go]: I wouldn't mind seeing an error thrown if a user tries to set `deleted` on create

File: website/docs/r/google_project_iam_role.html.markdown (hunk 2 above)
Comment (rosbo) [reply website/docs/r/google_project_iam_role.html.markdown]: I am ok with renaming the resource name. However, the `role_id` cannot have an hyphen: `It doesn't match pattern [a-zA-Z0-9_\.]{3,64}.`

File: website/docs/r/google_project_iam_role.html.markdown (hunk 3 above)
Comment (rosbo) [reply website/docs/r/google_project_iam_role.html.markdown]: Done

File: google/resource_google_project_iam_role_test.go (hunk 1 above)
Comment (rosbo) [reply google/resource_google_project_iam_role_test.go]: Done

File: google/resource_google_project_iam_role.go (hunk 4 above)
Comment (rosbo) [reply google/resource_google_project_iam_role.go]: Done

File: google/resource_google_project_iam_role.go (hunk 5 above)
Comment (rosbo) [reply google/resource_google_project_iam_role.go]: Done

**Comment by rosbo** [comment]
Yeah, I think also `_custom_role` is probably better since this creates and deletes custom roles. It doesn't deal with regular role. I think it is a good case to diverge from the API and use the nomenclature used in the GCP docs.

**Comment by rosbo** [comment]
I renamed the resource to `google_project_iam_custom_role`. I am also curious in @pdecat opinion on it. I can easily revert the latest comment.

**Comment by pdecat** [comment]
Hi @danawillow @rosbo, in the issue description, I just wanted to match the API as I thought was the norm.

I agree `google_project_iam_custom_role` is more explicit and was a bit confused at the beginning with the doc/API inconsistency.

Also, I guessed that if *custom* was not used in the API, it was probably to make it more generic and allow other usages in the future.

**Comment by rosbo** [comment]
Great, we are all in agreement :)

@danawillow could re-review it with the latest adjustments?

Thanks

**Pull request merged**